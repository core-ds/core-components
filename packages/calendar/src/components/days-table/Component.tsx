/* eslint-disable complexity */
import React, { FC, RefCallback, useCallback, useRef } from 'react';
import { CSSTransition,TransitionGroup } from 'react-transition-group';
import cn from 'classnames';
import isEqual from 'date-fns/isEqual';
import isLastDayOfMonth from 'date-fns/isLastDayOfMonth';
import isSameDay from 'date-fns/isSameDay';
import isToday from 'date-fns/isToday';
import isWithinInterval from 'date-fns/isWithinInterval';
import startOfMonth from 'date-fns/startOfMonth';

import { Button } from '@alfalab/core-components-button';
import { usePrevious } from '@alfalab/hooks';

import { Day } from '../../typings';
import { getSelectionRange, russianWeekDay,WEEKDAYS } from '../../utils';

import styles from './index.module.css';

export type DaysTableProps = {
    /**
     * Массив-календарь недель
     */
    weeks?: Day[][];

    /**
     * Активный месяц
     */
    activeMonth?: Date;

    /**
     * Начало выделенного периода
     */
    selectedFrom?: Date | number;

    /**
     * Конец выделенного периода
     */
    selectedTo?: Date | number;

    /**
     * Индикатор, что выбран полный период
     */
    rangeComplete?: boolean;

    /**
     * Подсвеченная дата (ховер)
     */
    highlighted?: Date | number;

    /**
     * Доп. пропсы для переданного дня
     */
    getDayProps: (
        day: Day,
    ) => Record<string, unknown> & {
        ref: RefCallback<HTMLTableDataCellElement>;
        onClick: (e: React.MouseEvent<HTMLTableDataCellElement>) => void;
    };

    /**
     * Нужно ли рендерить шапку
     */
    hasHeader?: boolean;

    /**
     * Должен ли календарь подстраиваться под ширину родителя.
     */
    responsive?: boolean;
};

export const DaysTable: FC<DaysTableProps> = ({
    weeks = [],
    activeMonth = new Date(),
    highlighted,
    selectedFrom,
    selectedTo,
    rangeComplete = selectedFrom && selectedTo,
    getDayProps,
    hasHeader = true,
    responsive,
}) => {
    const activeMonthRef = useRef(activeMonth);

    activeMonthRef.current = activeMonth;

    const prevActiveMonth = usePrevious(activeMonth);

    const direction = prevActiveMonth && (activeMonth < prevActiveMonth ? 'right' : 'left');

    const selection = getSelectionRange(selectedFrom, selectedTo, highlighted);

    const renderHeader = useCallback(
        () =>
            WEEKDAYS.map(dayName => (
                <th className={styles.dayName} key={dayName}>
                    {dayName}
                </th>
            )),
        [],
    );

    const renderDay = (day: Day, dayIdx: number) => {
        if (!day) return <td key={dayIdx} />;

        const daySelected =
            day.selected ||
            (selectedFrom && isSameDay(day.date, selectedFrom)) ||
            (selectedTo && isSameDay(day.date, selectedTo));

        const dayHighlighted = highlighted && isEqual(day.date, highlighted);
        const inRange = selection && isWithinInterval(day.date, selection);

        const firstDayOfMonth = day.date.getDate() === 1;
        const lastDayOfMonth = isLastDayOfMonth(day.date);

        const firstDayOfWeek = russianWeekDay(day.date) === 0;
        const lastDayOfWeek = russianWeekDay(day.date) === 6;

        const transitLeft = firstDayOfMonth && inRange && selection && day.date > selection.start;
        const transitRight = lastDayOfMonth && inRange && selection && day.date < selection.end;

        const rangeStart = selection && isSameDay(day.date, selection.start);
        const rangeEnd = selection && isSameDay(day.date, selection.end);

        const sharpTransitLeft =
            firstDayOfWeek &&
            firstDayOfMonth &&
            inRange &&
            selection &&
            (isSameDay(day.date, selection.start) || isSameDay(day.date, selection.end));

        const sharpTransitRight =
            lastDayOfWeek &&
            lastDayOfMonth &&
            inRange &&
            selection &&
            (isSameDay(day.date, selection.start) || isSameDay(day.date, selection.end));

        const dayProps = getDayProps(day);

        const { onClick } = dayProps;

        const handleDayClick = (e: React.MouseEvent<HTMLTableDataCellElement>) => {
            if (!day.disabled) onClick(e);
        };

        return (
            // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-noninteractive-element-interactions
            <td
                {...dayProps}
                key={day.date.getTime()}
                className={cn(styles.dayWrapper, {
                    [styles.range]: inRange,
                    [styles.rangeComplete]: inRange && rangeComplete,
                    [styles.transitLeft]: transitLeft,
                    [styles.transitRight]: transitRight,
                    [styles.sharpTransitLeft]: sharpTransitLeft,
                    [styles.sharpTransitRight]: sharpTransitRight,
                    [styles.rangeStart]: rangeStart,
                    [styles.rangeEnd]: rangeEnd,
                    [styles.cursorPointer]: !day.disabled,
                })}
                align='center'
                ref={node => {
                    /**
                     * После анимации реф-коллбэк вызывается еще раз, и в него передается null и старый activeMonth.
                     * Поэтому приходится хранить актуальный месяц в рефе и сравнивать с ним.
                     */
                    if (startOfMonth(day.date).getTime() === activeMonthRef.current.getTime()) {
                        dayProps.ref(node as HTMLTableDataCellElement);
                    }
                }}
                onClick={handleDayClick}
            >
                <Button
                    type='button'
                    view='ghost'
                    size='xs'
                    disabled={day.disabled}
                    className={cn(styles.day, {
                        [styles.selected]: daySelected,
                        [styles.today]: isToday(day.date),
                        [styles.disabled]: day.disabled,
                        [styles.holiday]: !day.disabled && day.holiday,
                        [styles.highlighted]: dayHighlighted,
                    })}
                >
                    {day.event && <span className={styles.dot} />}
                    {day.date.getDate()}
                </Button>
            </td>
        );
    };

    const renderWeek = (week: Day[], weekIdx: number) => (
        <tr key={weekIdx}>{week.map(renderDay)}</tr>
    );

    return (
        <table
            className={cn(styles.daysTable, direction && styles[direction], {
                [styles.responsive]: responsive,
            })}
        >
            {hasHeader && (
                <thead>
                    <tr>{renderHeader()}</tr>
                </thead>
            )}
            <TransitionGroup component={null}>
                <CSSTransition
                    key={activeMonth.getTime()}
                    timeout={300}
                    classNames={{
                        enter: styles.daysEnter,
                        enterActive: styles.daysEnterActive,
                        exit: styles.daysExit,
                        exitActive: styles.daysExitActive,
                    }}
                >
                    <tbody>{weeks.map(renderWeek)}</tbody>
                </CSSTransition>
            </TransitionGroup>
        </table>
    );
};
