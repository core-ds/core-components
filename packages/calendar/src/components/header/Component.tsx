import React, { FC, ReactNode } from 'react';
import cn from 'classnames';

import { SelectorView } from '../../typings';

import styles from './index.module.css';

export type HeaderProps = {
    /**
     * Вид шапки — месяц и год или только месяц
     */
    view?: SelectorView;

    /**
     * Отображать тень? (нужна при прокрутке)
     */
    withShadow?: boolean;

    /**
     * Дочерние элементы.
     */
    children?: ReactNode;
};

export const Header: FC<HeaderProps> = ({ view = 'full', withShadow, children }) => (
        <div
            className={cn(styles.header, {
                [styles.monthOnly]: view === 'month-only',
                [styles.withShadow]: withShadow,
            })}
            aria-live='polite'
        >
            {children}
        </div>
    );
