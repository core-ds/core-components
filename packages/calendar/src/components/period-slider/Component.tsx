import React, { FC, MouseEvent, useMemo } from 'react';
import cn from 'classnames';
import endOfWeek from 'date-fns/endOfWeek';
import startOfWeek from 'date-fns/startOfWeek';

import { Button } from '@alfalab/core-components-button';
import { IconButton } from '@alfalab/core-components-icon-button';
import { ChevronBackMIcon } from '@alfalab/icons-glyph/ChevronBackMIcon';

import { monthName } from '../../utils';

import { formatPeriod, getYearSelectorValue, shiftValues } from './utils';

import styles from './index.module.css';

export type PeriodType = 'range' | 'day' | 'week' | 'month' | 'quarter' | 'year';

export type PeriodSliderProps = {
    /**
     * Активная дата или период
     */
    value?: Date | [Date, Date];

    /**
     * Тип периода
     */
    periodType?: PeriodType;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Отключает кнопку назад
     */
    prevArrowDisabled?: boolean;

    /**
     * Отключает кнопку вперед
     */
    nextArrowDisabled?: boolean;

    /**
     * Скрывает заблокированные кнопки
     */
    hideDisabledArrows?: boolean;

    /**
     * Возможность выбора месяца и года, если periodType 'month'
     */
    isMonthAndYearSelectable?: boolean;

    /**
     * Отображать ли текущий год, если isMonthAndYearSelectable true
     */
    showCurrentYearSelector?: boolean;

    /**
     * Функция для форматирование выбранного периода
     */
    periodFormatter?: (valueFrom: Date, valueTo: Date, periodType: PeriodType) => string;

    /**
     * Обработчик нажатия кнопки переключения на назад
     */
    onPrevArrowClick?: (
        event: MouseEvent<HTMLButtonElement>,
        payload: {
            value: Date;
            valueFrom: Date;
            valueTo: Date;
            periodType: PeriodType;
        },
    ) => void;

    /**
     * Обработчик нажатия кнопки переключения на вперед
     */
    onNextArrowClick?: (
        event: MouseEvent<HTMLButtonElement>,
        payload: {
            value: Date;
            valueFrom: Date;
            valueTo: Date;
            periodType: PeriodType;
        },
    ) => void;

    /**
     * Обработчик нажатия на селектор месяца
     */
    onMonthClick?: (event: MouseEvent<HTMLButtonElement>) => void;

    /**
     * Обработчик нажатия на селектор года
     */
    onYearClick?: (event: MouseEvent<HTMLButtonElement>) => void;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;
};

export const PeriodSlider: FC<PeriodSliderProps> = ({
    value,
    periodType = 'month',
    className,
    periodFormatter = formatPeriod,
    prevArrowDisabled = false,
    nextArrowDisabled = false,
    hideDisabledArrows = false,
    isMonthAndYearSelectable = false,
    showCurrentYearSelector = false,
    onPrevArrowClick = () => null,
    onNextArrowClick = () => null,
    onMonthClick,
    onYearClick,
    dataTestId,
}) => {
    const [valueFrom, valueTo] = useMemo(() => {
        let from: Date;
        let to: Date;

        if (!value) return [undefined, undefined];

        if (Array.isArray(value)) {
            [from, to] = value;
        } else {
            [from, to] = [value, value];

            if (periodType === 'week') {
                from = startOfWeek(from, { weekStartsOn: 1 });
                to = endOfWeek(from, { weekStartsOn: 1 });
            }
        }

        return [from, to];
    }, [periodType, value]);

    const yearSelectorValue = useMemo(
        () => getYearSelectorValue(valueFrom, showCurrentYearSelector),
        [showCurrentYearSelector, valueFrom],
    );

    const showArrow = (direction: 'prev' | 'next') => {
        if (hideDisabledArrows) {
            const disabled = direction === 'prev' ? prevArrowDisabled : nextArrowDisabled;

            return !disabled && valueFrom;
        }

        return true;
    };

    const handleNextArrowClick = (event: MouseEvent<HTMLButtonElement>) => {
        if (!valueFrom || !valueTo) return;

        const newValues = shiftValues(valueFrom, valueTo, periodType, 'next');

        onNextArrowClick(event, {
            value: newValues.valueFrom,
            valueFrom: newValues.valueFrom,
            valueTo: newValues.valueTo,
            periodType,
        });
    };

    const handlePrevArrowClick = (event: MouseEvent<HTMLButtonElement>) => {
        if (!valueFrom || !valueTo) return;

        const newValues = shiftValues(valueFrom, valueTo, periodType, 'prev');

        onPrevArrowClick(event, {
            value: newValues.valueFrom,
            valueFrom: newValues.valueFrom,
            valueTo: newValues.valueTo,
            periodType,
        });
    };

    const renderHeader = () => {
        if (!(valueFrom && valueTo)) {
            return <span className={cn(styles.period, styles.empty)}>Укажите период</span>;
        }

        return periodType === 'month' && isMonthAndYearSelectable ? (
            <div>
                <Button className={styles.period} view='ghost' size='l' onClick={onMonthClick}>
                    {monthName(valueFrom)}
                </Button>
                {yearSelectorValue && (
                    <Button
                        className={cn(styles.yearSelectorButton, styles.period)}
                        view='ghost'
                        size='l'
                        onClick={onYearClick}
                    >
                        {yearSelectorValue}
                    </Button>
                )}
            </div>
        ) : (
            <span className={styles.period}>{periodFormatter(valueFrom, valueTo, periodType)}</span>
        );
    };

    return (
        <div
            className={cn(styles.component, className)}
            aria-live='polite'
            data-test-id={dataTestId}
        >
            {showArrow('prev') && (
                <IconButton
                    size='xs'
                    className={styles.arrow}
                    icon={ChevronBackMIcon}
                    onClick={handlePrevArrowClick}
                    disabled={prevArrowDisabled || !valueFrom}
                    aria-label='Предыдущий период'
                />
            )}

            {renderHeader()}

            {showArrow('next') && (
                <IconButton
                    size='xs'
                    className={styles.arrow}
                    icon={ChevronBackMIcon}
                    onClick={handleNextArrowClick}
                    disabled={nextArrowDisabled || !valueFrom}
                    aria-label='Следующий период'
                />
            )}
        </div>
    );
};
