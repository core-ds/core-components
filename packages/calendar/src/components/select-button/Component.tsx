import React, { ButtonHTMLAttributes, forwardRef } from 'react';
import cn from 'classnames';

import { ButtonDesktop as Button } from '@alfalab/core-components-button/desktop';

import styles from './index.module.css';

export type SelectButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Вид кнопки
     */
    view?: 'default' | 'filled' | 'current' | 'selected';
};

export const SelectButton = forwardRef<HTMLButtonElement, SelectButtonProps>(
    ({ className, children, view = 'default', ...restProps }, ref) => (
        <Button
            {...restProps}
            ref={ref}
            view='text'
            size='xs'
            className={cn(styles.button, styles[view], className)}
        >
            {children}
        </Button>
    ),
);
