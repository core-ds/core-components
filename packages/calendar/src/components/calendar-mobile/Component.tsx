import React, { forwardRef, useMemo, useRef } from 'react';
import { Virtuoso } from 'react-virtuoso';
import { ResizeObserver as ResizeObserverPolyfill } from '@juggle/resize-observer';
import cn from 'classnames';
import endOfDay from 'date-fns/endOfDay';
import startOfDay from 'date-fns/startOfDay';
import startOfMonth from 'date-fns/startOfMonth';

import { Button } from '@alfalab/core-components-button';
import { ModalMobile } from '@alfalab/core-components-modal/mobile';

import { Calendar, CalendarProps, limitDate, monthName, useCalendar, WEEKDAYS } from '../..';
import { Month } from '../../typings';
import { dateArrayToHashTable, generateMonths, generateWeeks } from '../../utils';
import { DaysTable } from '../days-table';

import styles from './index.module.css';

// ResizeObserverPolyfill необходим для корректной работы react-virtuoso.
if (typeof window !== 'undefined' && !window.ResizeObserver) {
    window.ResizeObserver = ResizeObserverPolyfill;
}

export type CalendarMobileProps = CalendarProps & {
    /**
     * Управление видимостью модалки
     */
    open: boolean;

    /**
     * Заголовок календаря
     */
    title?: string;

    /**
     * Обработчик закрытия модалки
     */
    onClose?: () => void;

    /**
     * Количество лет для генерации в обе стороны от текущего года
     */
    yearsAmount?: number;

    /**
     * Нужно ли рендерить шапку
     */
    hasHeader?: boolean;

    /**
     * Разрешить выбор из недозаполненного диапазона дат.
     */
    allowSelectionFromEmptyRange?: boolean;
};

const CalendarMonthOnlyView = ({
    value,
    defaultView,
    month: monthTimestamp,
    minDate: minDateTimestamp,
    maxDate: maxDateTimestamp,
    defaultMonth: defaultMonthTimestamp,
    offDays,
    events,
    holidays,
    onChange,
    selectedFrom,
    selectedTo,
    rangeComplete,
    onMonthChange,
    yearsAmount = 3,
}: CalendarMobileProps) => {
    const initialMonthIndex = useMemo(() => {
        const currentMonthIndex = new Date().getMonth();

        return yearsAmount * 12 + currentMonthIndex;
    }, [yearsAmount]);

    const month = useMemo(
        () => (monthTimestamp ? new Date(monthTimestamp) : undefined),
        [monthTimestamp],
    );

    const minDate = useMemo(
        () => (minDateTimestamp ? startOfDay(minDateTimestamp) : undefined),
        [minDateTimestamp],
    );

    const maxDate = useMemo(
        () => (maxDateTimestamp ? endOfDay(maxDateTimestamp) : undefined),
        [maxDateTimestamp],
    );

    const selected = useMemo(() => (value ? new Date(value) : undefined), [value]);

    const defaultMonth = useMemo(
        () =>
            startOfMonth(
                selected ||
                    limitDate(
                        defaultMonthTimestamp || Date.now(),
                        minDateTimestamp,
                        maxDateTimestamp,
                    ),
            ),
        [defaultMonthTimestamp, maxDateTimestamp, minDateTimestamp, selected],
    );

    const { activeMonth, months, highlighted, getDayProps } = useCalendar({
        month,
        defaultMonth,
        view: defaultView,
        minDate,
        maxDate,
        selected,
        offDays,
        events,
        onChange,
        onMonthChange,
    });

    const activeMonths = useMemo(() => {
        const eventsMap = dateArrayToHashTable(events || []);
        const offDaysMap = dateArrayToHashTable(offDays || []);
        const holidaysMap = dateArrayToHashTable(holidays || []);
        const prevMonths: Month[] = [];
        const nextMonths: Month[] = [];

        const date = new Date();
        const currentYear = date.getFullYear();

        for (let i = 0; i < yearsAmount; i++) {
            const prevYear = date.setFullYear(currentYear - (i + 1));
            const nextYear = date.setFullYear(currentYear + (i + 1));

            const prevYearMonths = generateMonths(new Date(prevYear), {});
            const nextYearMonths = generateMonths(new Date(nextYear), {});

            prevMonths.unshift(...prevYearMonths);
            nextMonths.push(...nextYearMonths);
        }

        const generatedMonths = [...prevMonths, ...months, ...nextMonths];

        return generatedMonths.map((item) => ({
            ...item,
            weeks: generateWeeks(item.date, {
                minDate,
                maxDate,
                selected,
                eventsMap,
                offDaysMap,
                holidaysMap,
            }),
            title: `${monthName(item.date)} ${item.date.getFullYear()}`,
        }));
    }, [events, offDays, holidays, months, yearsAmount, minDate, maxDate, selected]);

    const renderMonth = (index: number) => (
        <div className={styles.daysTable} id={`month-${index}`}>
            <span className={styles.month}>{activeMonths[index].title}</span>
            <DaysTable
                weeks={activeMonths[index].weeks}
                activeMonth={activeMonth}
                selectedFrom={selectedFrom}
                selectedTo={selectedTo}
                getDayProps={getDayProps}
                highlighted={highlighted}
                rangeComplete={rangeComplete}
                hasHeader={false}
                responsive={true}
            />
        </div>
    );

    return (
        <Virtuoso
            totalCount={activeMonths.length}
            itemContent={renderMonth}
            initialTopMostItemIndex={initialMonthIndex}
            increaseViewportBy={800}
            itemSize={(el) => el.getBoundingClientRect().height + 32}
        />
    );
};

export const CalendarMobile = forwardRef<HTMLDivElement, CalendarMobileProps>(
    (
        {
            hasHeader = true,
            allowSelectionFromEmptyRange = false,
            className,
            defaultView = 'days',
            selectorView = 'full',
            value,
            selectedFrom,
            selectedTo,
            onChange,
            dataTestId,
            open,
            onClose,
            title = 'Календарь',
            yearsAmount = 3,
            ...restProps
        },
        ref,
    ) => {
        const modalRef = useRef<HTMLDivElement>(null);

        const monthOnlyView = selectorView === 'month-only';

        const handleClose = () => {
            if (onClose) onClose();
        };

        const handleClear = () => {
            if (onChange) onChange();
        };

        const renderDayNames = () => (
            <table className={styles.dayNames}>
                <thead>
                    <tr>
                        {WEEKDAYS.map((dayName) => (
                            <th className={styles.dayName} key={dayName}>
                                {dayName}
                            </th>
                        ))}
                    </tr>
                </thead>
            </table>
        );

        const renderContent = () => {
            const commonProps = {
                value,
                onChange,
                defaultView,
                selectorView,
                selectedFrom,
                selectedTo,
            };

            if (monthOnlyView) {
                return (
                    <CalendarMonthOnlyView
                        open={open}
                        yearsAmount={yearsAmount}
                        {...commonProps}
                        {...restProps}
                    />
                );
            }

            return (
                <Calendar
                    responsive={true}
                    className={styles.calendar}
                    {...commonProps}
                    {...restProps}
                />
            );
        };

        const renderFooter = () => {
            if (selectedFrom || selectedTo) {
                let selectButtonDisabled = !selectedFrom || !selectedTo;

                if (allowSelectionFromEmptyRange) {
                    selectButtonDisabled = !selectedFrom;
                }

                return (
                    <React.Fragment>
                        <Button
                            view='primary'
                            size='s'
                            block={true}
                            onClick={handleClose}
                            disabled={selectButtonDisabled}
                        >
                            Выбрать
                        </Button>
                        <Button view='secondary' size='s' block={true} onClick={handleClear}>
                            Сбросить
                        </Button>
                    </React.Fragment>
                );
            }

            if (value) {
                return (
                    <Button view='primary' size='s' block={true} onClick={handleClose}>
                        Выбрать
                    </Button>
                );
            }

            return (
                <Button view='secondary' size='s' block={true} onClick={handleClose}>
                    Отмена
                </Button>
            );
        };

        return (
            <div className={cn(className, styles.component)} ref={ref} data-test-id={dataTestId}>
                <ModalMobile
                    open={open}
                    onClose={handleClose}
                    ref={modalRef}
                    className={styles.modal}
                    wrapperClassName={styles.wrapper}
                >
                    {hasHeader && (
                        <ModalMobile.Header
                            hasCloser={true}
                            title={title}
                            align='center'
                            leftAddons={<div />}
                            sticky={true}
                            className={cn({ [styles.withZIndex]: selectorView === 'full' })}
                        />
                    )}
                    {monthOnlyView && renderDayNames()}
                    <ModalMobile.Content flex={true}>{renderContent()}</ModalMobile.Content>
                    <ModalMobile.Footer
                        sticky={true}
                        className={cn({ [styles.withZIndex]: selectorView === 'full' })}
                    >
                        {renderFooter()}
                    </ModalMobile.Footer>
                </ModalMobile>
            </div>
        );
    },
);
