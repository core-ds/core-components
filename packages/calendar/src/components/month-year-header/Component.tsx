import React, { FC, MouseEvent } from 'react';
import cn from 'classnames';

import { ChevronDownCompactSIcon } from '@alfalab/icons-glyph/ChevronDownCompactSIcon';

import { monthName } from '../../utils';
import { SelectButton } from '../select-button';

import styles from './index.module.css';
import mobileStyles from './mobile.module.css';

export type MonthYearHeaderProps = {
    /**
     * Активная дата
     */
    value?: Date;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Обработчик нажатия на кнопку месяца
     */
    onMonthClick?: (event: MouseEvent<HTMLButtonElement>) => void;

    /**
     * Обработчик нажатия на кнопку года
     */
    onYearClick?: (event: MouseEvent<HTMLButtonElement>) => void;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;

    /**
     * Мобильная версия компонента
     */
    mobile?: boolean;
};

export const MonthYearHeader: FC<MonthYearHeaderProps> = ({
    value,
    className,
    onMonthClick,
    onYearClick,
    dataTestId,
    mobile,
}) => {
    const month = value ? monthName(value) : undefined;
    const year = value ? value.getFullYear().toString() : undefined;

    return (
        <div
            className={cn(styles.component, className, {
                [mobileStyles.component]: mobile,
            })}
            aria-live='polite'
            data-test-id={dataTestId}
        >
            <SelectButton
                view='filled'
                className={cn(styles.button, styles.month, {
                    [mobileStyles.button]: mobile,
                })}
                onClick={onMonthClick}
            >
                <span className={styles.buttonContent}>
                    {month}
                    <ChevronDownCompactSIcon className={styles.upDownIcon} />
                </span>
            </SelectButton>
            <SelectButton
                view='filled'
                className={cn(styles.button, styles.year, {
                    [mobileStyles.button]: mobile,
                })}
                onClick={onYearClick}
            >
                <span className={styles.buttonContent}>
                    {year}
                    <ChevronDownCompactSIcon className={styles.upDownIcon} />
                </span>
            </SelectButton>
        </div>
    );
};
