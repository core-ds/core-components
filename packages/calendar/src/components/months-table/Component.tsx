import React, { FC, useCallback } from 'react';
import cn from 'classnames';
import isSameMonth from 'date-fns/isSameMonth';
import isThisMonth from 'date-fns/isThisMonth';

import { Month } from '../../typings';
import { monthName } from '../../utils';
import { SelectButton, SelectButtonProps } from '../select-button';

import styles from './index.module.css';

export type MonthsTableProps = {
    /**
     * Массив месяцев
     */
    months?: Month[];

    /**
     * Выбранный месяц
     */
    selectedMonth?: Date;

    /**
     * Доп пропсы для переданного месяца
     */
    getMonthProps: (day: Month) => Record<string, unknown>;

    /**
     * Должен ли календарь подстраиваться под ширину родителя.
     */
    responsive?: boolean;
};

export const MonthsTable: FC<MonthsTableProps> = ({
    selectedMonth,
    months = [],
    getMonthProps,
    responsive,
}) => {
    const view = useCallback(
        (month: Month): SelectButtonProps['view'] => {
            if (selectedMonth && isSameMonth(selectedMonth, month.date)) return 'selected';
            if (isThisMonth(month.date)) return 'outlined';

            return 'default';
        },
        [selectedMonth],
    );

    return (
        <div className={cn(styles.monthsTable, { [styles.responsive]: responsive })}>
            {months.map(month => (
                <SelectButton
                    {...getMonthProps(month)}
                    key={month.date.getTime()}
                    className={styles.button}
                    view={view(month)}
                >
                    {monthName(month.date)}
                </SelectButton>
            ))}
        </div>
    );
};
