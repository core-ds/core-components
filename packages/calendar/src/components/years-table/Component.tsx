import React, { FC, MouseEvent, useCallback, useLayoutEffect, useRef } from 'react';
import cn from 'classnames';
import isSameYear from 'date-fns/isSameYear';
import isThisYear from 'date-fns/isThisYear';

import { SelectButton, SelectButtonProps } from '../select-button';

import styles from './index.module.css';

export type YearsTableProps = {
    /**
     * Массив лет
     */
    years?: Date[];

    /**
     * Выбранный год
     */
    selectedYear?: Date;

    /**
     * Доп. пропсы для переданного года
     */
    getYearProps: (year: Date) => Record<string, unknown>;

    /**
     * Обработчик скролла
     */
    onScroll: (scrollTop: number) => void;

    /**
     * Должен ли календарь подстраиваться под ширину родителя.
     */
    responsive?: boolean;
};

export const YearsTable: FC<YearsTableProps> = ({
    selectedYear,
    years = [],
    getYearProps,
    onScroll,
    responsive,
}) => {
    const ref = useRef<HTMLDivElement>(null);

    const view = useCallback(
        (year: Date): SelectButtonProps['view'] => {
            if (selectedYear && isSameYear(selectedYear, year)) return 'selected';
            if (isThisYear(year)) return 'outlined';

            return 'default';
        },
        [selectedYear],
    );

    const handleScroll = useCallback(
        (event: MouseEvent<HTMLDivElement>) => {
            onScroll(event.currentTarget.scrollTop);
        },
        [onScroll],
    );

    useLayoutEffect(() => {
        const listNode = ref.current;
        const selector = `.${styles.button}[tabIndex="0"]`;
        const selectedYearNode = listNode && listNode.querySelector<HTMLButtonElement>(selector);

        if (listNode && selectedYearNode) {
            const topIndent = listNode.clientHeight / 2 - selectedYearNode.clientHeight / 2;

            listNode.scrollTop = selectedYearNode.offsetTop - topIndent;

            onScroll(listNode.scrollTop);
        }
    }, [onScroll, selectedYear]);

    return (
        <div
            className={cn(styles.yearsTable, { [styles.responsive]: responsive })}
            onScroll={handleScroll}
            ref={ref}
        >
            <div className={styles.inner}>
                {years.map(year => (
                    <SelectButton
                        {...getYearProps(year)}
                        key={year.getFullYear()}
                        view={view(year)}
                        className={styles.button}
                    >
                        {year.getFullYear()}
                    </SelectButton>
                ))}
            </div>
        </div>
    );
};
