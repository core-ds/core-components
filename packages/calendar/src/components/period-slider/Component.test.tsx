import React from 'react';
import { render, screen } from '@testing-library/react';
import { PeriodSlider } from './Component';

describe('PeriodSlider', () => {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
            matches: true,
            media: query,
            onchange: null,
            addListener: jest.fn(), // Deprecated
            removeListener: jest.fn(), // Deprecated
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });

    describe('empty PeriodSlider', () => {
        it('empty PeriodSlider has default title', () => {
            render(<PeriodSlider periodType='day' />);
            const defaultTitle = screen.queryByText('Укажите период');
            expect(defaultTitle).toBeInTheDocument();
        });

        it('empty PeriodSlider has specific title', () => {
            render(<PeriodSlider periodType='day' emptyValueText='Не выбрано' />);
            const specificTitle = screen.queryByText('Не выбрано');
            expect(specificTitle).toBeInTheDocument();
        });

        it('empty PeriodSlider default title snapshot', () => {
            const { container } = render(<PeriodSlider periodType='day' />);
            expect(container).toMatchSnapshot();
        });

        it('empty PeriodSlider specific title snapshot', () => {
            const { container } = render(
                <PeriodSlider periodType='day' emptyValueText='Не выбрано' />,
            );
            expect(container).toMatchSnapshot();
        });
    });
});
