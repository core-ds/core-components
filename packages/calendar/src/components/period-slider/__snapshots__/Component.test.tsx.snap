// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeriodSlider empty PeriodSlider empty PeriodSlider default title snapshot 1`] = `
<div>
  <div
    aria-live="polite"
    class="component"
  >
    <button
      aria-label="Предыдущий период"
      class="component text size-48 hug size-48 component text component text cc-icon-button arrow primary component"
      disabled=""
      type="button"
    >
      <span
        class="label stretchText"
      >
        <span
          class="iconWrapper size-32 center"
        >
          <svg
            class="icon"
            fill="currentColor"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M10 12l5.5 5.5L14 19l-7-7 7-7 1.5 1.5L10 12z"
            />
          </svg>
        </span>
      </span>
    </button>
    <span
      class="period empty"
    >
      Укажите период
    </span>
    <button
      aria-label="Следующий период"
      class="component text size-48 hug size-48 component text component text cc-icon-button arrow primary component"
      disabled=""
      type="button"
    >
      <span
        class="label stretchText"
      >
        <span
          class="iconWrapper size-32 center"
        >
          <svg
            class="icon"
            fill="currentColor"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M10 12l5.5 5.5L14 19l-7-7 7-7 1.5 1.5L10 12z"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`PeriodSlider empty PeriodSlider empty PeriodSlider specific title snapshot 1`] = `
<div>
  <div
    aria-live="polite"
    class="component"
  >
    <button
      aria-label="Предыдущий период"
      class="component text size-48 hug size-48 component text component text cc-icon-button arrow primary component"
      disabled=""
      type="button"
    >
      <span
        class="label stretchText"
      >
        <span
          class="iconWrapper size-32 center"
        >
          <svg
            class="icon"
            fill="currentColor"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M10 12l5.5 5.5L14 19l-7-7 7-7 1.5 1.5L10 12z"
            />
          </svg>
        </span>
      </span>
    </button>
    <span
      class="period empty"
    >
      Не выбрано
    </span>
    <button
      aria-label="Следующий период"
      class="component text size-48 hug size-48 component text component text cc-icon-button arrow primary component"
      disabled=""
      type="button"
    >
      <span
        class="label stretchText"
      >
        <span
          class="iconWrapper size-32 center"
        >
          <svg
            class="icon"
            fill="currentColor"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M10 12l5.5 5.5L14 19l-7-7 7-7 1.5 1.5L10 12z"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</div>
`;
